backend/migrations/dataall_migrations/migrationmanager.py

```diff
@@ -0,0 +1,124 @@
+            return True
+
+        revision = self.current_migration.next_migration()
+        while revision is not None:
+            if revision.revision_id() == target_revision_id:
+                return True
+            revision = revision.next_migration()
+
+        Exception(f'Failed to find {target_revision_id}.')
+
+    def upgrade(self, target_revision_id=None):
+        if not self.check_upgrade_id(target_revision_id):
+            logger.info('Data-all version is up to date')
+            return self.current_migration.revision_id()
+
+        logger.info(f"Upgrade from {self.current_migration.revision_id()} to {target_revision_id or 'latest'}")
+        executed_upgrades: Deque[Union[Type[MigrationBase], None]] = deque()
+        saved_start_migration = self.current_migration
+        while self.current_migration is not None:
+            try:
+                logger.info(f'Applying migration {self.current_migration.__name__}')
+                self.current_migration.up()
+                executed_upgrades.append(self.current_migration)
+                logger.info(f'Migration {self.current_migration.__name__} completed')
+                if (
+                    self.current_migration.revision_id() == target_revision_id
+                    or self.current_migration.next_migration() is None
+                ):
+                    break
+                self.current_migration = self.current_migration.next_migration()
+            except Exception as e:
+                # Something went wrong revert
+                logger.info(f'An error occurred while applying the migration.{e}.')
+                while executed_upgrades:
+                    migration = executed_upgrades.pop()
+                    migration.down()
+                self.current_migration = saved_start_migration
+                return False
+        logger.info('Upgrade completed')
+        self.save_upgraded(executed_upgrades)
+        return self.current_migration.revision_id()
+
+    def downgrade(self, target_revision_id='0'):
+        if not self.check_downgrade_id(target_revision_id):
+            logger.info(f'Current revision is  {self.current_migration.revision_id()}')
+            return
+
+        logger.info(f"Downgrade from {self.current_migration.revision_id()} to {target_revision_id or 'initial'}")
+        executed_downgrades: Deque[Union[Type[MigrationBase], None]] = deque()
+        while self.current_migration:
+            if self.current_migration.revision_id() == target_revision_id:
+                break
+            try:
+                logger.info(f'Reverting migration {self.current_migration.__name__}')
+                self.current_migration.down()
+                executed_downgrades.append(self.current_migration)
+                logger.info(f'Migration {self.current_migration.__name__} completed')
+            except Exception as e:
+                logger.info(f'An error occurred while reverting the migration.{e}.')
```

Comment:
I wonder if that's something we want to do (upgrade again in case downgrade fail). Also similar to my upgrade function should we call `self.upgrade` here (but avoid inf loops)?